File: set6/6/gh_sol.py
dp = [0] * 20002
n, M = map(int, input().split(' '))
dp[0] = dp[2] = 1
for i in range(4, n, 2):
    dp[i] = dp[i-2]
    for j in range(2, i, 2):
        dp[i] = dp[i] + dp[j] * dp[i-j-2]
    dp[i] %= M
print(dp[n-2])

File: set6/6/gh_sol.java
import java.util.*;
class Main
{
    public static void main(String[] args)
    {
        Scanner sc = new Scanner(System.in);
        String []token = sc.nextLine().split(" ");
        int n = Integer.parseInt(token[0]);
        int M = Integer.parseInt(token[1]);
        long []dp = new long[4002];
        dp[0] = dp[2] = 1L;
        for(int i=4; i<=4000; i=i+2)
        {
            dp[i] = dp[i-2];
            for(int j=2; j<i; j=j+2)
            {
                dp[i] = dp[i] + dp[j]*dp[i-j-2];
                dp[i] %= M;
            }
        }
        System.out.println(dp[n-2]+"");
    }
}
